### PR TITLE
Remove -bounds-checking-unique-traps (replace with -fno-sanitize-merge=local-bounds)

### PR DESCRIPTION
**Pull request by thurstond**
#120613 removed -ubsan-unique-traps and replaced it with  -fno-sanitize-merge (introduced in #120511), which allows fine-grained control of which UBSan checks to prevent merging. This analogous patch removes -bound-checking-unique-traps, and allows it to be controlled via -fno-sanitize-merge=local-bounds.

Most of this patch is simply plumbing through the compiler flags into the bounds checking pass.

Note: this patch subtly changes -fsanitize-merge (the default) to also include -fsanitize-merge=local-bounds. This is different from the previous behavior, where -fsanitize-merge (or the old -ubsan-unique-traps) did not affect local-bounds (requiring the separate -bounds-checking-unique-traps). However, we argue that the new behavior is more intuitive.

Removing -bounds-checking-unique-traps and merging its functionality into -fsanitize-merge breaks backwards compatibility; we hope that this is acceptable since '-mllvm -bounds-checking-unique-traps' was an experimental flag.

### CODE REVIEW

File: clang/docs/ReleaseNotes.rst [hunk 1]
```diff
@@ -488,8 +489,11 @@ Removed Compiler Flags
   derivatives) is now removed, since it's no longer possible to suppress the
   diagnostic (see above). Users can expect an `unknown warning` diagnostic if
   it's still in use.
-- The experimental flag '-ubsan-unique-traps' has been removed. It is
-  superseded by '-fno-sanitize-merge'.
+- The experimental flags '-ubsan-unique-traps' and
```
Comment (vitalybuka) [inline clang/docs/ReleaseNotes.rst:492]: this is -mllvm flag, they don't worth mentioning

File: clang/docs/ReleaseNotes.rst [hunk 2]
```diff
@@ -445,9 +445,10 @@ New Compiler Flags
 - The ``-Warray-compare-cxx26`` warning has been added to warn about array comparison
   starting from C++26, this warning is enabled as an error by default.
 
-- '-fsanitize-merge' (default) and '-fno-sanitize-merge' have been added for
-  fine-grained control of which UBSan checks are allowed to be merged by the
-  backend (for example, -fno-sanitize-merge=bool,enum).
+- ``-fsanitize-merge`` (default) and ``-fno-sanitize-merge`` have been added for
```
Comment (vitalybuka) [inline clang/docs/ReleaseNotes.rst:448]: there is a special section "Sanitizers" on line 1154, maybe better to keep there?

File: llvm/test/Instrumentation/BoundsChecking/runtimes.ll [hunk 3]
```diff
@@ -5,6 +5,21 @@
 ; RUN: opt < %s -passes='bounds-checking<rt-abort>'     -S | FileCheck %s --check-prefixes=RTABORT
 ; RUN: opt < %s -passes='bounds-checking<min-rt>'       -S | FileCheck %s --check-prefixes=MINRT
 ; RUN: opt < %s -passes='bounds-checking<min-rt-abort>' -S | FileCheck %s --check-prefixes=MINRTABORT
+;
+; merge defaults to true
+; RUN: opt < %s -passes='bounds-checking<merge=true>'              -S | FileCheck %s --check-prefixes=TR
```
Comment (vitalybuka) [inline llvm/test/Instrumentation/BoundsChecking/runtimes.ll:10]: I think one with trap and one with runtime is enough
we will need later one flag for guard, we don't want 3x them

File: llvm/lib/Passes/PassBuilder.cpp [hunk 4]
```diff
@@ -1281,31 +1281,43 @@ parseRegAllocFastPassOptions(PassBuilder &PB, StringRef Params) {
...
-      Mode = BoundsCheckingPass::ReportingMode::FullRuntimeAbort;
+      Options.Mode = BoundsCheckingPass::ReportingMode::FullRuntimeAbort;
     } else if (ParamName == "min-rt") {
-      Mode = BoundsCheckingPass::ReportingMode::MinRuntime;
+      Options.Mode = BoundsCheckingPass::ReportingMode::MinRuntime;
     } else if (ParamName == "min-rt-abort") {
-      Mode = BoundsCheckingPass::ReportingMode::MinRuntimeAbort;
+      Options.Mode = BoundsCheckingPass::ReportingMode::MinRuntimeAbort;
+    } else if (ParamName.consume_front("merge=")) {
+      if (ParamName == "true")
```
Comment (vitalybuka) [inline llvm/lib/Passes/PassBuilder.cpp:1302]: we don't do true/false here

usually if it's missing -> false, if it's present -> true

Check `ParamName == "kernel"` in this file

File: llvm/lib/Transforms/Instrumentation/BoundsChecking.cpp [hunk 5]
```diff
@@ -299,7 +303,7 @@ void BoundsCheckingPass::printPipeline(
     raw_ostream &OS, function_ref<StringRef(StringRef)> MapClassName2PassName) {
   static_cast<PassInfoMixin<BoundsCheckingPass> *>(this)->printPipeline(
       OS, MapClassName2PassName);
-  switch (Mode) {
+  switch (Options.Mode) {
```
Comment (vitalybuka) [inline llvm/lib/Transforms/Instrumentation/BoundsChecking.cpp:306]: you need to update this one function as well, this is opposite of parse

**Review by vitalybuka — APPROVED** [review]

File: clang/docs/ReleaseNotes.rst (hunk 1 above)
Comment (thurstond) [reply clang/docs/ReleaseNotes.rst]: Removed

File: clang/docs/ReleaseNotes.rst (hunk 2 above)
Comment (thurstond) [reply clang/docs/ReleaseNotes.rst]: Moved

File: llvm/test/Instrumentation/BoundsChecking/runtimes.ll (hunk 3 above)
Comment (thurstond) [reply llvm/test/Instrumentation/BoundsChecking/runtimes.ll]: Ok, trimmed the tests

File: llvm/lib/Passes/PassBuilder.cpp (hunk 4 above)
Comment (thurstond) [reply llvm/lib/Passes/PassBuilder.cpp]: Changed to simple match of 'merge'

File: clang/docs/ReleaseNotes.rst [hunk 6]
```diff
@@ -1208,6 +1202,11 @@ Sanitizers
 
 - Implemented ``-f[no-]sanitize-trap=local-bounds``, and ``-f[no-]sanitize-recover=local-bounds``.
 
+- ``-fsanitize-merge`` (default) and ``-fno-sanitize-merge`` have been added for
+  fine-grained, unified control of which UBSan checks can potentially be merged
+  by the compiler (for example,
```
Comment (vitalybuka) [inline clang/docs/ReleaseNotes.rst:1207]: Please mention on  https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html#stack-traces-and-report-symbolization as well

**Review by vitalybuka — APPROVED** [review]

**Pull request merged**